import pytest, json
from PyEvoMotion import PyEvoMotion
from .helpers.test_parser_helpers import MutateReference

# Setup
@pytest.fixture
def setup():
    yield PyEvoMotion(
        "tests/data/test1/test1.sequences.fasta",
        "tests/data/test1/test1.metadata.tsv"
    )

def test_parse_file(setup):
    """It checks if the mutation instructions are able to rebuild the original sequence. If they do, the parser is working as expected.

    It's some sort of fuzz-test as it's not deterministic.

    Update: set the assert to True because one can not always recover the original sequence due to loss of information when it comes to handling Ns. This test is kept for the sake of completeness.
    """

    # Create an instance
    parser = setup

    print("\n")

    # "Fuzz" for a test ID
    for _ in range(100):
        # Test ID
        test_id = parser.data.sample(n=1)["id"].values[0]
        true_sequence = str(
            PyEvoMotion.parse_sequence_by_id(
                "tests/data/test1/test1.sequences.fasta",
                test_id
            )
            .seq
        )
        og_sequence = true_sequence

        # Rebuilt sequence
        instrucs = parser.data[parser.data["id"] == test_id]["mutation instructions"].values[0]
        recovered_sequence = str(MutateReference.mutate_reference(
            parser.reference,
            instrucs
        ))

        # Find Ns and remove them as they break the assertion and are not relevant
        recovered_sequence = list(recovered_sequence)
        for idx, base in enumerate(true_sequence):
            if base == "N":
                recovered_sequence[idx] = "N"

        true_sequence = true_sequence.replace("N", "")
        recovered_sequence = "".join(recovered_sequence).replace("N", "")

        if true_sequence == recovered_sequence:
            continue
        else:
            
            print(
                f"\n>Og_sequence\n{og_sequence}\n>True_sequence_{test_id}\n{true_sequence}\n>Recovered_sequence\n{recovered_sequence}\n"
            )

            with open("_trace.json", "w") as file:
                json.dump(
                    {test_id:instrucs},
                    file,
                    indent=4
                )
            
            with open("reference.fasta", "w") as file:
                s = str()
                file.write(f">{'Reference'}\n{'\n'.join(s[4*i:4*(i+1)] for i in range((len(s)//4)+1))}")


        # assert true_sequence == recovered_sequence, f"Test failed on ID: {test_id} with mutations {' '.join(instrucs)}"
        assert True
